### ISSUE TITLE
Depend on django_component Component

### ISSUE THREAD

**Issue by EmilStenstrom** [tetra-framework/tetra#78]
**Breaking this out into a separate issue as to not confuse the original thread:**

Originally posted by @nerdoc in https://github.com/tetra-framework/tetra/issues/74#issuecomment-2439435717

@EmilStenstrom Ah, ok, that's a much better approach with .py as entry file, and right, renaming a component should be made easy, and the less files you rename, the better. 

My thoughts about merging tetra/django_components (which is already a bit off-topic):
* I think, especially in open source projects, working together is *always* better than the currently hyped business model "My NIH-project is better than all the others". So I like the idea very much to work together as close as possible, as much of the code is somehow duplicated, at least in function. There are a few things to consider. And a few ways how to combine them:
  * Tetra could e.g. replace its BasicComponent with DC, and add additional functionality - the `Component` API on top of it. I think this approach is feasible and wouldn't even change Tetra's API. I like the "library" concept of Tetra very much, although the API is a bit cumbersome. Putting components into directories would solve the "library" approach too - Tetra's internal usage of the Library class could even be kept, the whole thing just needs refactoring then.
  * Tetra could merge with django_components, one project is abandoned. Hm, this is maybe not the best way to deal with it.

@EmilStenstrom how do you think about it? I always thought in a way that duplicating work in open source projects is a BadIdea™. It would be some effort to align Tetra's and DC's aims, but I think it would be worth it. It would also allow users to use a framework of their choice (DC for basic "stacking" of components within HTML, Tetra+DC for additional "reactiveness").

There could be other benefits, as Tetra is very opinonated with its way of holding state, there may be other on-top frameworks emerging that keep state on the server etc., all basing on DC. DC is very flexible ATM regarding configurability. I for myself like the approach better to use one path that works for 95% of the users, and works well, and leave not too much open for "cosmetic" configuration, as it basically makes codebases that use different approaches (`{% component "foo_bar" ... %}`, `{% FooBar ...%}`) at least *optically* incompatible or incoherent. But flexibility has advantages too.

But it would mean we'd have to work together more closely and help out each other maybe (which, for my part, is perfectly ok).

...my 2 cents.

EDIT: another thing - I wouldn't omit @samwillis ' opinion here - as he originally created this framework, and I don't want to wreak havok on the codebase - merging with/basing upon Django Components would at least break a few things - starting from the `js`/`script` component's attributes, up to different ways of naming slots/blocks within the templates, variable scopes etc.
So Sam, could you tell us maybe what you think about this?

_Originally posted by @nerdoc in https://github.com/tetra-framework/tetra/issues/74#issuecomment-2439435717_

**Comment by EmilStenstrom** [comment]
I think both projects are interesting for different kinds of projects, so merging them is probably not the right path. 
* Tetra delivers an opinionated, reactive, "full project" experience
* Django_components delivers building blocks, we even have a long-term goal of merging with Django

The value proposal for you depending on django_components would be that you get access to all the features of django_components, but you would have to "pay" for that with refactoring time, adapting Tetra to work with django_components. This is of course solely a decision that you can make, I can just say that we we would be open to discussing how we can help to make that happen.

I'm guessing the question of opinionatedness would come up there. Were you want to add something to django_components that we would feel is not a great fit? On the other hand, extending the Component class is easy, so you would be able to adapt it that way. 

I'm not sure what the best way of exploring this would be?

**Comment by EmilStenstrom** [comment]
@nerdoc Fully agree that we shouldn't merge, and django_components has evolved a lot since this thread was started (thanks to @JuroOravec). I would say we're more of a "component fundamentals" framework now. Would love to have you switch tetra to be built on top of djc, and we would be happy to help with that migration if you want!

**Comment by JuroOravec** [comment]
Thinking a bit more about this, what would actually make more sense would be to first integrate Tetra with DJC as an [extension](https://django-components.github.io/django-components/latest/concepts/advanced/extensions/).

The problem with my previous comment was that it would mean:
- Long time spent refactoring
- Breaking peoples projects

I still think it makes sense in the long run for Tetra to get rid of the component & templating logic and delegate that to DJC. But it should be done gradually.

If if made that extension, Tetra users would have plenty of time to gradually migrate from using Tetra components to DJC components. And one wouldn't have to wait until we implement the missing features in DJC.

The integration could look like this:

1. User install Django, django-components, tetra
2. User adds django-components and tetra to INSTALLED_APPS
3. User DOESN'T have to set Tetra's middleware nor urlpatterns. Instead, they configure just django-components'
4. User installs the integration (e.g. djc-ext-tetra), and adds it to django-components' settings.

**Comment by JuroOravec** [comment]
Thinking a bit more about this, what would actually make more sense would be to first integrate Tetra with DJC as an [extension](https://django-components.github.io/django-components/latest/concepts/advanced/extensions/).

The problem with my previous comment was that it would mean:
- Long time spent refactoring
- Breaking peoples projects

I still think it makes sense in the long run for Tetra to get rid of the component & templating logic and delegate that to DJC. But it should be done gradually.

If we made that as a DJC extension, Tetra users would have plenty of time to gradually migrate from using Tetra components to DJC components. And one wouldn't have to wait until we implement the missing features in DJC.

The integration could look like this:

1. User install Django, django-components, tetra
2. User adds django-components and tetra to INSTALLED_APPS
3. User DOESN'T have to set Tetra's middleware nor urlpatterns. Instead, they configure just django-components
4. User installs the integration (e.g. djc-ext-tetra), and adds it to django-components' settings.

    ```py
    # settings.py

    COMPONENTS = ComponentsSettings(
        extensions=[
            "djc_tetra.TetraExtension",
        ],
    )
    ```
5. When writing up a Component, the Tetra-specific variables and methods would be scoped under `Tetra` nested class:

    ```py
    # django-components part
    class ToDoList(Component):
        template = """
          <div>
             ...
          </div>
        """

        # Tetra part
        class Tetra:
            title = public("")
        
            def load(self, *args, **kwargs):
                self.todos = ToDo.objects.filter(session_key=self.request.session.session_key)

            @public
            def add_todo(self, title):
                todo = ToDo(
                    title=title,
                    session_key=self.request.session.session_key,
                )
                todo.save()
                self.title = ""
    ```

I think the extension would be also much easier to implement. Everything under DJC's `Component.Tetra` would be equivalent to Tetra's `Component` class. So we would just need to refactor that (`components/base.py` file) so that the `Component` class can be also used standalone (without Tetra's `BaseComponent`).

**Comment by EmilStenstrom** [comment]
> Hello [@JuroOravec](https://github.com/JuroOravec) again. I like the idea of depending on djc more and more. I read a bit of the docs on djc extensions, but I don't get it really: Someone can make extensions that can be "plugged" into individual DJC components. But is is it possible to make something like a "base class" like `TetraComponent" or "ReactiveComponent" in a djc-tetra module that can be used like Tetra's Component class? I mean, I'd like to "subclass" a DJC `Component`, name it `ReactiveComponent`or`TetraComponent` - and put all the logic of tetra into it.
> 
> But I can't find a way in the Extensions to do that. Am I mising something

Juro have to correct me if I’m wrong here, but I would expect Tetra to be more than an DJC extension. Instead, the base class you are asking for would be something that lives inside of tetra, and that you get when you import tetra in your project.

You could of course register extensions as well, but I would expect the majority of the code to be inbyte subclass.

**Comment by JuroOravec** [comment]
And to reply to older messages:

> Hello @JuroOravec again. I like the idea of depending on djc more and more. I read a bit of the docs on djc extensions, but I don't get it really:
> Someone can make extensions that can be "plugged" into individual DJC components.
> But is is it possible to make something like a "base class" like TetraComponent" or "ReactiveComponent" in a djc-tetra module that can be used like Tetra's Component class? I mean, I'd like to "subclass" a DJC Component, name it ReactiveComponentorTetraComponent` - and put all the logic of tetra into it.

Extensions and subclasses are 2 separate approaches
- If we were to go with extensions approach, the DJC's Component class would remain as is. Instead, Tetra logic would have to be scoped under `Component.Tetra`.
- Could could also achieve what you're suggest with subclassing. *However*, DJC components use metaclasses. So instead of subclassing a `Component` class, we'd have  to subclass the `ComponentMeta` metaclass, probably in Tetra's [`BasicComponentMetaClass`](https://github.com/tetra-framework/tetra/blob/3f9faeaf5af277b27fd420f02d972856e39287e0/tetra/components/base.py#L192):

But as I said in previous message, I think actually a better way would be to not inherit from DJC directly, and instead have the DJC class under `_component_cls` attribute. As this would allow to have control over what fields / behaviour is exposed and which is not.

----

As for the earlier msg from [here](https://github.com/tetra-framework/tetra/issues/78#issuecomment-3058498775):

> You mean it would be possible:
>
> {% dynamic is=dyn_component %}{% enddynamic %}
> Where dyn_component is a variable with a name.

Yes!

**Comment by nerdoc** [comment]
Hmmmm.  Yes, there is a render() method in Tetra components. But the method needs the current component's data (properties, like `title`, `done`, `todo` in the `Todo` class), because it needs to render them. So A Tetra component has to "override" the render method anyway, or DJC must expose some render hooks where tetra can hook into (which would be the better approach).In DJC this does not matter, as components have no individual state. But to allow later sub-projects a per-component state, the render() method must change anyway IMHO. or at least allow "somehow" rendering a state in a instance method that is called. But then the render() method must get the instance as argument and call it's modifier method.

OTOH, I like the idea of having Tetra components as facade pattern - so the namespace is clean for properties, and the DJC class is internally handled, with rendering delegated. This is brilliant - I'll try to accomplish that - just need to see how this performs.

Regarding the subclass/extension discussion: I must admit that I do not fully understand the extension system yet. But it seems to be more complicated than I first thought. And the documentation is not too verbose here...
Maybe you could give me a hint. Do I need a "TetraExtension" that is included then in settings:

```python
COMPONENTS = ComponentsSettings(
    extensions=[
        TetraExtension,
    ],
)
```
And then ALL components are using this extension? Or just the ones that define a

This must be done by the user then, ant can't be included "automatically" when using tetra? Maybe you provide a backend hook as well, so an app could use an extension "as dependency".

What I really want, is to keep the syntax for users simple. I'd definitely go the way of reducing the boilerplate code as much as possible, e.g. I'd like to remove the "@" or "component" tag all together, using DJC's approach of automatically registering the tags (like `{% calendar ... / %}` - I'd like to implement this next in Tetra to converge the Tetra/DJC syntax further - but Tetra will definitely be more opiniated than DJC in terms of syntax.

**Comment by JuroOravec** [comment]
### Passing Tetra state to DJC component

> Yes, there is a render() method in Tetra components. But the method needs the current component's data (properties, like title, done, todo in the Todo class), because it needs to render them. So A Tetra component has to "override" the render method anyway, or DJC must expose some render hooks where tetra can hook into (which would be the better approach)

Ah, so if I understand right, this is about the fact that:
- DJC is stateless, so the component's only input is what the user passes to it.
- ~Tetra has state, so there's 2 sources of inputs - user input and the current state.~ Actually, component input IS turned into component state.
- ~And so there should be a way to merge the two when calling DJC component from within Tetra component.~

Update: Actually looking at the example on Tetra's loading page. It seems that Tetra's flow is:
  - When component is instantiated: Component input -> Component state
  - And when it comes to rendering the component: Component state -> Render context.

Component state is added to the context in [`_add_self_attrs_to_context`](https://github.com/tetra-framework/tetra/blob/c6da98ac254a8c255a89f831ab5c15fa2846bcf6/tetra/components/base.py#L390C14-L390C40)

But the logic in `_add_self_attrs_to_context()` looks like it will work also with a Dict without having to change anything.

So for example in Tetra's `BaseComponent.render()` we could do something like this:

```py
def render(self) -> SafeString:
    # Same as before
    # NOTE: `get_context_data` does NOT need to add component state to the Context.
    #              This should be done by DJC when we pass the component state as kwargs.
    self.recalculate_attrs(component_method_finished=True)
    context = self.get_context_data()

    # NEW - Create dict with component state
    kwargs = self._add_self_attrs_to_context({})

    # NOTE: I removed the code block that manages Django's {% block %} tags.
    #             I think it can be skipped, should be handled by DJC.

    # Call underlying DJC component
    html = self._component_cls.render(context, kwargs=kwargs)

    # NOTE: mark_safe() shouldn't be needed with DJC, but not 100% sure.
    return mark_safe(html)
```

Sidenote: DJC does allow a hook to inject /  modify component inputs (see [`on_component_input`](https://django-components.github.io/django-components/latest/reference/extension_hooks/#django_components.extension.ComponentExtension.on_component_input)). But as shown, I don't think we need it in this case.

### Keeping Tetra interface clean

> OTOH, I like the idea of having Tetra components as facade pattern - so the namespace is clean for properties, and the DJC class is internally handled, with rendering delegated. This is brilliant - I'll try to accomplish that - just need to see how this performs.

Sounds great 🔥

And yeah, I can imagine it will require a lot of playing around. Happy to give it test when you've got some working version.

-----

### Managing DJC settings via Tetra

> Do I need a "TetraExtension" that is included then in settings?
>
> ...
>
> This must be done by the user then, ant can't be included "automatically" when using tetra? Maybe you provide a backend hook as well, so an app could use an extension "as dependency".

The way I see it, it's about who is in control / who is the source of truth:

- If Tetra was a DJC extension, it's DJC that's the main framework and source of truth, so it IMO makes sense to define settings for it explicitly.

- If Tetra is using DJC internally, then yes, it would make sense to define all the settings via Tetra, and then Tetra passes those settings to DJC.
   - As  an example, Tetra could have a `TETRA_COMPONENTS` field in `settings.py`, which would serve as a replacement for DJC's `COMPONENTS` field.

   - For just a proof of concept, I think we can just use the `COMPONENTS` field directly.
      
     But as I eventually want to make DJC independent of Django, I'm also thinking about how to be able to configure the source of settings.

     I don't know the details yet, but the idea is that you would be somehow able to override [this line](https://github.com/JuroOravec/django-components/blob/595ca8c5f279e40207d7b02d34ccbde35a942d81/src/django_components/app_settings.py#L753) that decides where the settings are loaded from.

     So at that point, Tetra would be able to inject itself into the settings loading logic:

    ```py
    # tetra/components.py
    from .extentension import TetraExtension

    # Load settings from Django's settings.py, but:
    # 1. Using `TETRA_COMPONENTS` instead of `COMPONENTS`
    # 2. And injecting Tetra as extension
    def load_components_settings():
        from django.conf import settings
        data = getattr(settings, "COMPONENTS", {})

        # Normalize to ComponentsSettings so we know that the `extensions`
        # field exists
        components_settings = (
            ComponentsSettings(**data)
            if not isinstance(data, ComponentsSettings)
            else data
        )

        # Add Tetra as extension if not already
        if TetraExtension not in components_settings.extensions:
            components_settings.append(TetraExtension)
        
    # NOTE: Once DJC is separated from Django,
    # you will have to create a global `Components` instance
    # (from which you will access the public API)
    components = Components(
        load_settings=load_components_settings,
    )
    ```

----

### Are extensions enabled for all components?

> And then ALL components are using this extension? Or just the ones that define a Tetra subclass? I have to learn more here...

Right now how it works is that the extension itself decides for which Components it will be enabled.

For example both the caching extension and the View extension have an opt-in field that the user has to set to `True` in order to activate that extension.

With Tetra, I think you could go a step further and make the extension so that it automatically opts itself in when it sees that there is at least one state set for the component.

So in other words:

This would automatically enable Tetra for this component class:

```py
from django_components import Component

class Table(Component):
    class Tetra:
        title = public("")
        done = public(False)

        @public.watch("title", "done")
        @public.debounce(200)
        def save(self, value, old_value, attr):
            self.todo.title = self.title
            self.todo.done = self.done
            self.todo.save()
```

But not in this:

```py
class Table(Component):
    # Ignored - No public state
    class Tetra:
        pass
```

And not in this:

```py
# Again, no public Tetra state defined
class Table(Component):
    pass
```
